test/success/hlevels.prl

```diff
@@ -0,0 +1,68 @@
+Def HasAllPathsTo (#C,#c) = [(-> [c' : #C] (path [_] #C c' #c))].
+Def IsContr (#C) = [(* [c : #C] (HasAllPathsTo #C c))].
+Def IsProp (#A) = [(-> [a b : #A] (path [_] #A a b))].
+Def IsSet (#A) = [(-> [a b : #A] (IsProp (path [_] #A a b)))].
+
+Thm InhPropIsContr(#l:lvl) : [
+ (->
+  [ty : (U #l kan)] // can I use pre here?
+  [h : (IsProp ty)]
+  [a : ty]
+  (IsContr ty))
+] by [
+  unfold IsProp;
+  lam ty, h, a. {use a,lam x. `($ h x a)}
+].
+
+Thm PropPi(#l:lvl) : [
+ (->
+  [tyA : (U #l kan)]
+  [tyB : (-> tyA (U #l kan))]
+  [h : (-> [x : tyA] (IsProp ($ tyB x)))]
+  (IsProp (-> [x : tyA] ($ tyB x))))
+] by [
+  unfold IsProp; lam tyA, tyB, h, f, g.
+  <i> lam x. `(@ ($ h x ($ f x) ($ g x)) i)
+].
+
+Thm PropSet(#l:lvl) : [
+  (->
+   [tyA : (U #l kan)]
+   [h : (IsProp tyA)]
+   (IsSet tyA)
+  )
+] by [
+  unfold [IsSet, IsProp];
+  lam tyA, h, a, b, p, q.
+  <j i> `(hcom 0~>1 tyA a
+               [i=0 [k] (@ ($ h a a) k)]
+               [i=1 [k] (@ ($ h a b) k)]
+               [j=0 [k] (@ ($ h a (@ p i)) k)]
+               [j=1 [k] (@ ($ h a (@ q i)) k)])
+].
+
+// This proof should be easy, but it is not...
```

Comment:
So this is still quite complicated... Any ideas how to make it better or why it is so complicated? It is essentially just a direct application of  `PropSet`...

Reply:
I just checked this. Ideally the proof should be done by
```
lam tyA, h1, h2. <i> lam a, b. `(@ ($ (PropSet #l) tyA h1 a b ($ h1 a b) ($ h2 a b)) i)
```
but the auto-unfolding in the matching judgments is not implemented yet. :cry: Note the differences between parameters to meta-variables and arguments to functions. @jonsterling Perhaps we can treat all the extra parameters as arguments, but at very least some error messages are needed.

Reply:
As I described before, it's not likely that writing out the proof term like this is going to work... You really do have to interact with the refiner by using tactics, and that's a blessing rather than a curse: it lets you develop the proof interactively, rather than by having to come up with this proof term all by yourself. If you look at the proof that's written down, it's not really that much more complicated than the proof term you wanted; the only part about it which is more complicated is the fact that you have to unfold things here and there, which is a shame. We may be able to improve that...

Also, regarding auto-unfolding in those match judgments (a technical issue that @favonia will understand), if we want to do this, it should be done in the part of the `auto` tactic that deals with `MATCH` judgments, rather than in the rules for those things. Does that sound reasonable?

Reply:
I am wondering if @mortberg refers to `IsPropIsProp`. If so, in that special, simple case, the proof term should work once we have auto-unfolding and maybe other minor improvements.

Reply:
Yes, I am referring to `IsPropIsProp`! It seems to me like RedPRL is going to be extremely difficult to use if just applying a simple lemma turns into 7 lines of code... Ideally the proof should be exactly like what @favonia wrote. So making the automation handle the unfolding better seems like something to prioritize.

Reply:
I too am referring to `IsPropIsProp`. I think that auto-unfolding in the appropriate places will help, as I said before. But it is not at all the case that the ideal code is what @favonia wrote; we do *not* want users to be typing in terms manually (this is too low-level), they really should be using the interactive tactics. What we want instead is to get unfolding to work well-enough in practice that the tactics are not longer the manual codes.

Here's how it might look if unfolding worked better:

```
  lam tyA, h1, h2.
  <i> lam a, b. use (PropSet #l) [`tyA, `h1, `a, `b, use h1 [`a, `b], use h2 [`a, `b], `i]
```

The above is the ideal code. (Obviously, the concrete notation leaves something to be desired, but what I have written above has the structure of the proof that we want; just quoting a term and hoping that we can guess a typing derivation for it is not what we want.)

Reply:
One thing I want to stress is that the current `auto-step` tactic is poor enough that there are a lot of low-hanging fruit that can improve the current situation (see issue #438 for @favonia's thoughts on it). But it is not a problem that we can solve in the general case, and it's likely that doing "easy" things in RedPRL will *nearly always* be harder than doing "easy" things in a system with decidable judgmental equality.

This is the trade-off, and we can negotiate it to the point of making many things nicer than they are, but it will always be an issue. We're hoping that the gain in expressiveness which we get will make it worth it, but that is by no means guaranteed. These kinds of difficulties are sadly part of the territory of implementing "extensional type theory" and/or "computational type theory", and we can never really overcome them in a real sense.

Reply:
I see! I guess I'm just too used to writing proof terms directly these days :)

How about merging this PR now and then I'll update it once the ideal code works?